### PR TITLE
expose more detailed signature errors

### PR DESCRIPTION
**Pull request by hannesm**
also, even if hash_whitelist contains a weak algorithm, print a log message
(warning level) about the usage of the weak algorithm

//cc @cfcs -- we discussed this

### CODE REVIEW

File: lib/validation.ml [hunk 1]
```diff
@@ -11,17 +45,35 @@ let issuer_matches_subject
...
 
-let validate_raw_signature hash_whitelist raw signature_algo signature_val pk_info =
+let validate_raw_signature subject hash_whitelist raw signature_algo signature_val pk_info =
   match pk_info, Algorithm.to_signature_algorithm signature_algo with
   | `RSA issuing_key, Some (`RSA, siga) ->
-    List.mem siga hash_whitelist &&
-    ( match Mirage_crypto_pk.Rsa.PKCS1.sig_decode ~key:issuing_key signature_val with
-      | None           -> false
+    begin match
+        Mirage_crypto_pk.Rsa.PKCS1.sig_decode ~key:issuing_key signature_val
```
[the remark this answers is not in the document]
Comment (hannesm) [reply lib/validation.ml]: excellent idea, done in 20e0992

File: lib/x509.mli [hunk 2]
```diff
@@ -443,6 +449,196 @@ module Certificate : sig
...
+      fingerprint list (for a trust on first use implementation).
+  *)
+
+  (** The type of signature verification errors. *)
+  type signature_error = [
+    | `Bad_pkcs1_signature of Distinguished_name.t * Cstruct.t
+    | `Hash_algorithm_mismatch of Distinguished_name.t * Mirage_crypto.Hash.hash * Mirage_crypto.Hash.hash
+    | `Hash_mismatch of Distinguished_name.t * Cstruct.t * Cstruct.t
+    | `Hash_not_whitelisted of Distinguished_name.t * Mirage_crypto.Hash.hash
+    | `Bad_pkcs1_digest_info of Distinguished_name.t * Cstruct.t
```
[the remark this answers is not in the document]
Comment (hannesm) [reply lib/x509.mli]: done in 37f9bfb

File: lib/x509.mli [hunk 3]
```diff
@@ -443,6 +449,196 @@ module Certificate : sig
...
+  (** {1 Fingerprint verification} *)
+
+  (** [trust_key_fingerprint host ~time ~hash ~fingerprints certificates] is
+      [result], the first element of [certificates] is verified against the
+      given [fingerprints] list using {!key_fingerprint}.  If [time] is
+      provided, the certificate has to be valid at the given timestamp.  If
+      [host] is provided, the certificate is checked for this name.  The
+      [hostname] of the fingerprint list must match the name in the certificate,
+      using {!hostnames}. *)
+  val trust_key_fingerprint :
```
[the remark this answers is not in the document]
Comment (hannesm) [reply lib/x509.mli:616]: note: this is not new code, only needed to be shuffled around since Validation.signature_error is used in multiple modules.

> extract cert + pubkey fingerprints from certs

isn't [tlsclient](https://github.com/hannesm/tlsclient) the right place for this? it already prints the certificate and public key fingerprint (of the chain).

> does the need for [`host] Domain_name.t prevent using this for *.foo.com wildcard certificates for HTTPS use?

TL;DR: you provide a ``[`host] Domain_name.t * Cstruct.t`` list, and if the server certificate matches on of the Cstruct.t values, it's names (common name / subject alternative name) are checked to include the `` [`host] Domain_name.t `` -- i.e. you specify ("foo.example.com", "00:11:22:33") as authenticator, then if the server certificate fingerprint matches "00:11:22:33", its names are validated against "foo.example.com" -- if it has a subjectAlternativeName with "*.example.com", this is fine :)

[the remark this answers is not in the document]
Comment (cfcs) [reply lib/x509.mli:616]: I think `tlsclient` sounds like a good place. I think people who are looking at this mli string would appreciate a pointer to that. :)

**Comment by cfcs** [comment]
I think this PR looks good. :)
